validator/client/src/main/java/tech/pegasys/teku/validator/client/restapi/apis/PostVoluntaryExit.java

```diff
@@ -58,17 +84,79 @@ public PostVoluntaryExit() {
             .pathParam(PARAM_PUBKEY_TYPE)
             .queryParam(EPOCH_PARAMETER)
             .description(
-                "Create a signed voluntary exit message for an active validator, identified by a public key known to the validator client. This endpoint returns a SignedVoluntaryExit object, which can be used to initiate voluntary exit via the beacon node's submitPoolVoluntaryExit endpoint.")
+                "Create a signed voluntary exit message for an active validator, identified by a public key known "
+                    + "to the validator client. This endpoint returns a SignedVoluntaryExit object, which can be "
+                    + "used to initiate voluntary exit via the beacon node's submitPoolVoluntaryExit endpoint.")
             .response(SC_OK, "Success response", SIGNED_VOLUNTARY_EXIT_RESPONSE_TYPE)
             .withNotFoundResponse()
             .withUnauthorizedResponse()
             .withForbiddenResponse()
-            .withNotImplementedResponse()
             .build());
+    this.spec = spec;
+    this.keyManager = keyManager;
+    this.validatorApiChannel = validatorApiChannel;
+    this.timeProvider = timeProvider;
   }
 
   @Override
   public void handleRequest(RestApiRequest request) throws JsonProcessingException {
-    request.respondError(SC_NOT_IMPLEMENTED, "Not implemented");
+    final BLSPublicKey publicKey = request.getPathParameter(PARAM_PUBKEY_TYPE);
+    final UInt64 epoch =
+        request
+            .getOptionalQueryParameter(EPOCH_PARAMETER)
+            .orElseGet(
+                () -> {
+                  final SpecVersion genesisSpec = spec.getGenesisSpec();
+                  final UInt64 genesisTime = genesisSpec.getConfig().getMinGenesisTime();
+                  final UInt64 currentTime = timeProvider.getTimeInSeconds();
+                  final UInt64 slot =
+                      genesisSpec.miscHelpers().computeSlotAtTime(genesisTime, currentTime);
+                  return spec.computeEpochAtSlot(slot);
+                });
+
+    final SafeFuture<SignedVoluntaryExit> future = getVoluntaryExit(publicKey, epoch);
+    request.respondAsync(future.thenApply(AsyncApiResponse::respondOk));
+  }
+
+  private SafeFuture<SignedVoluntaryExit> getVoluntaryExit(
+      final BLSPublicKey publicKey, final UInt64 epoch) {
+    return validatorApiChannel
```

Comment:
NIT: consider breaking some of this logic into smaller private methods to make this easier to read.

Reply:
better still, package private that you can test if appropriate :) indices map and how that's handled is a good example of a level where we could feed data in rather than wait for futures etc...